src/online2bin/online2-net-nnet3-latgen-faster.cc

```diff
@@ -0,0 +1,398 @@
+
+    feature_opts.Register(&po);
+    decodable_opts.Register(&po);
+    decoder_opts.Register(&po);
+    endpoint_opts.Register(&po);
+
+    po.Read(argc, argv);
+
+    if (po.NumArgs() != 4) {
+      po.PrintUsage();
+      return 1;
+    }
+
+    std::string nnet3_rxfilename = po.GetArg(1),
+        fst_rxfilename = po.GetArg(2),
+        word_syms_rxfilename = po.GetArg(3),
+        port_rspecifier = po.GetArg(4);
+
+    int port_num = std::stoi(port_rspecifier);
+
+    OnlineNnet2FeaturePipelineInfo feature_info(feature_opts);
+
+    KALDI_LOG << "Loading AM...";
+
+    TransitionModel trans_model;
+    nnet3::AmNnetSimple am_nnet;
+    {
+      bool binary;
+      Input ki(nnet3_rxfilename, &binary);
+      trans_model.Read(ki.Stream(), binary);
+      am_nnet.Read(ki.Stream(), binary);
+      SetBatchnormTestMode(true, &(am_nnet.GetNnet()));
+      SetDropoutTestMode(true, &(am_nnet.GetNnet()));
+      nnet3::CollapseModel(nnet3::CollapseModelConfig(), &(am_nnet.GetNnet()));
+    }
+
+    // this object contains precomputed stuff that is used by all decodable
+    // objects.  It takes a pointer to am_nnet because if it has iVectors it has
+    // to modify the nnet to accept iVectors at intervals.
+    nnet3::DecodableNnetSimpleLoopedInfo decodable_info(decodable_opts,
+                                                        &am_nnet);
+
+    KALDI_LOG << "Loading FST...";
+
+    fst::Fst<fst::StdArc> *decode_fst = ReadFstKaldiGeneric(fst_rxfilename);
+
+    fst::SymbolTable *word_syms = NULL;
+    if (!word_syms_rxfilename.empty())
+      if (!(word_syms = fst::SymbolTable::ReadText(word_syms_rxfilename)))
+        KALDI_ERR << "Could not read symbol table from file "
+                  << word_syms_rxfilename;
+
+    KALDI_LOG << "Loaded evertyhing. Waiting for data...";
+
+    TcpServer server;
+
+    server.Listen(port_num);
+
+    while (true) {
+
+      server.Accept();
+
+      int32 samp_count = 0;
+      size_t chunk_len = static_cast<size_t>(chunk_length_secs * samp_freq);
+
+      int32 check_update = static_cast<int32>(samp_freq * output_freq);
+      int32 next_check = check_update;
+
+      bool eos = false;
+
+      OnlineIvectorExtractorAdaptationState adaptation_state(
+          feature_info.ivector_extractor_info);
+
+      while (!eos) {
+
+        OnlineNnet2FeaturePipeline feature_pipeline(feature_info);
+        feature_pipeline.SetAdaptationState(adaptation_state);
+
+        OnlineSilenceWeighting silence_weighting(
+            trans_model,
+            feature_info.silence_weighting_config,
+            decodable_opts.frame_subsampling_factor);
+
+        SingleUtteranceNnet3Decoder decoder(decoder_opts, trans_model,
+                                            decodable_info,
+                                            *decode_fst, &feature_pipeline);
+
+        std::vector<std::pair<int32, BaseFloat> > delta_weights;
+
+        while (true) {
+
+          eos = !server.ReadChunk(chunk_len);
```

Comment:
Are you sure the eos part works with `nc`? AFAIK `nc` won't send a TCP half-shutdown, so you'll never get eos unless you stop `nc` in which case the client won't receive the final result (i.e. L248) because it's disconnected.

Reply:
Well, the samples I added in the docs worked fine. Regardless if the connection was interrupted abruptly (ctrl-c) or simply reached the end of file, the client (nc) disconnected cleanly and waited to accept the next connection. Did you have any problem running the samples?

Upon further inspection, the ctrl-c case does timeout for a few seconds, but it does go on waiting for the next client after that.

Reply:
Still trying to figure this out. Select doesn't seem to detect this CTRL-c scenario. Changing socket read timeout or keep alive timeout doesn't help either. My Linux simply goes through select, blocks in read for ~5 seconds and exists. Not sure what would unblock it sooner.

You think this should be really implemented in multiple threads? Isn't there a better option?

Reply:
Ok, I fixed it!

I looked at the sourcecode of netcat and seems poll is better for this kind of scenario than select. I went with poll for now and it works a treat.